lib/core/checker/name_resolution_checker.cc

```diff
@@ -0,0 +1,230 @@
+    errors_.push_back({"Enum's name " + ast_enum.name.GetString() + " is already in scope."});
+  }
+  (*this)(ast_enum.type_dependencies);
+
+  (*this)(ast_enum.pattern_mapping);
+  PopScope();
+}
+
+void NameResolutionChecker::operator()(const std::vector<ast::Enum::Rule> &rules) {
+  for (const auto &rule : rules) {
+    (*this)(rule);
+  }
+}
+
+void NameResolutionChecker::operator()(const ast::Enum::Rule &rule) {
+  PushScope();
+
+  isShadowing_ = true;
+
+  (*this)(rule.inputs);
+
+  isShadowing_ = false;
+
+  (*this)(rule.outputs);
+
+  PopScope();
+}
+
+void NameResolutionChecker::operator()(std::vector<ast::Constructor> &constructors) {
+  for (const auto &constructor : constructors) {
+    (*this)(constructor);
+  }
+}
+
+void NameResolutionChecker::operator()(const ast::Constructor &constructor) {
+  if (IsInScope(constructor.name)) {
+    errors_.push_back(
+        {"Constructor's name " + constructor.name.GetString() + " is already in scope."});
+  }
+  (*this)(constructor.fields);
+}
+
+void NameResolutionChecker::operator()(const ast::Star &) {}
+
+void NameResolutionChecker::operator()(
+    const std::unordered_map<InternedString, ast::Message> &messages) {
+  for (const auto &message : messages) {
+    (*this)(message.second);
+  }
+}
+
+void NameResolutionChecker::operator()(const ast::Message &ast_message) {
+  PushScope();
+  if (IsInScope(ast_message.name)) {
+    errors_.push_back({"Message's name " + ast_message.name.GetString() + " is already in scope."});
+  }
+  (*this)(ast_message.type_dependencies);
+  (*this)(ast_message.fields);
+  PopScope();
+}
+
+void NameResolutionChecker::operator()(const std::vector<ast::TypedVariable> &dependencies) {
+  for (const auto &dependency : dependencies) {
+    (*this)(dependency);
+  }
+}
+
+void NameResolutionChecker::operator()(const ast::TypedVariable &typed_variable) {
+  (*this)(typed_variable.type_expression);
+  AddName(typed_variable.name);
+}
+
+void NameResolutionChecker::operator()(const ast::TypeExpression &type_expression) {
+  if (!IsInScope(type_expression.name)) {
+    errors_.push_back({"Type's name " + type_expression.name.GetString() + " is not in scope."});
+  }
+
+  (*this)(type_expression.parameters);
+}
+
+void NameResolutionChecker::operator()(
+    const std::vector<std::unique_ptr<ast::Expression>> &expressions) {
+  for (const auto &expression : expressions) {
+    std::visit(*this, *expression);
+  }
+}
+
+void NameResolutionChecker::operator()(const ast::ConstructedValue &value) {
+  if (!IsInScope(value.constructor_identifier)) {
+    errors_.push_back(
+        {"Constructor's identifier " + value.constructor_identifier.GetString() +
+         " is not in scope."});
+  }
+  PushScope();
+  (*this)(value.constructor_identifier, value.fields);
+  PopScope();
+}
+
+void NameResolutionChecker::operator()(
+    InternedString constructor_identifier,
+    const std::vector<std::pair<InternedString, std::unique_ptr<ast::Expression>>> &fields) {
```

Comment:
This one gave me the idea to make fields inside a constructor typed variables. I don't know how I missed that